12747 [a11y] Make sure IDs are unique

bkjohnson opened the pull request:
## Description

Fixes https://github.com/department-of-veterans-affairs/va.gov-team/issues/12747

When forms allow multiple "items" to be added, there exists a problem where the `id` attribute for each added item would be the same, causing an accessibility issue for screen readers.

This PR changes it so that when successive items are added, their `id` in the DOM is appended with an integer corresponding to their order.


## Testing done

Local browser testing & unit testing


## Screenshots

Here you can see a `1` at the end of the id, which is there because the first item in the military service history list has a `0` appended to the end.  This pattern avoids the accessibility issue.

![image](https://user-images.githubusercontent.com/2008881/91781927-7e107280-ebb0-11ea-931b-2053a9f1088d.png)


## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
Append an int to the end of ids in ArrayField

This needs to be done for both the schema and the uiSchema

bkjohnson committed:
Remove useless const

bkjohnson committed:
Remove console logs

bkjohnson committed:
Try to fix broken tests

Add a default & do some null checks

bkjohnson committed:
Add some docs for the helper functions

bkjohnson committed:
Remove unnecessary const

We don't need each item to have the index at the end - we just need the
uiSchema and schema objects to be numerated correctly.

bkjohnson commented on src/platform/forms-system/src/js/review/ArrayField.jsx:
```diff
@@ -223,9 +285,14 @@ class ArrayField extends React.Component {
               !itemCountLocked &&
               (!schema.minItems || items.length > schema.minItems);
-            const itemSchema = this.getItemSchema(index);
+            const itemSchema = getNumeratedItemSchema(this.props.schema, index);
```
This ensures that the `schema` is numerated appropriately.

bkjohnson commented on src/platform/forms-system/src/js/review/ArrayField.jsx:
```diff
@@ -223,9 +285,14 @@ class ArrayField extends React.Component {
+            const itemSchema = getNumeratedItemSchema(this.props.schema, index);
             const itemTitle = itemSchema ? itemSchema.title : '';
 
+            const numeratedUiSchema = getNumeratedUiSchema(
+              uiSchema.items,
+              index,
+            );
```
And this ensures that the `uiSchema` is numerated appropriately.

bkjohnson committed:
Move helper functions out of class definition

These seem like they belong elsewhere

bkjohnson committed:
Fix import

bkjohnson committed:
Add tests & fix bug in code which used wrong index

bkjohnson commented on src/platform/forms-system/src/js/helpers.js:
```diff
@@ -642,3 +644,53 @@ export function transformForSubmit(
+ */
+export function getNumeratedItemSchema(schema, index) {
+  if (schema.items.length > index) {
+    return schema.items[index];
```
I may need to enumerate things here as well :thinking:

bkjohnson committed:
Replace getItemSchema class method with a helper

I'm not super confident with this yet - will do some testing

billfienberg commented on src/platform/forms-system/test/js/helpers.unit.spec.js:
```diff
@@ -1003,4 +1005,98 @@ describe('Schemaform helpers:', () => {
+    };
+
+    it('should numerate the keys for uiSchema fields', () => {
+      const numeratedUiSchema = getNumeratedUiSchema(uiSchema, 4);
```
`4` is an arbitrary `index`, correct?

bkjohnson replied:
Yup, it's arbitrary. When this function gets called in a `map` callback though, each successive call will have an index incremented by 1.

cvalarida commented on src/platform/forms-system/src/js/helpers.js:
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
 }
+
+/**
+ * Numerate the field names (keys) in a uiSchema object.
```
Probably worth calling out in the description here that it intentionally avoids adding numbers to keys which begin with `ui:`.

cvalarida commented on src/platform/forms-system/src/js/helpers.js:
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
+ * Numerate the field names (keys) in a uiSchema object.
+ *
+ * @param {Object} uiSchema - The uiSchema for the field
+ * @param {number} index = The integer to append to the field names
```
```suggestion
 * @param {number} index - The integer to append to the field names
```

cvalarida commented on src/platform/forms-system/src/js/helpers.js:
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
+}
+
+/**
+ * Almost a copy of the class method `getItemSchema`.
```
Does this refer to the `ArrayField` class method which you're deleting?

bkjohnson replied:
It does - that's some leftover copypasta.  I'll update accordingly.

cvalarida reviewed (commented):
The only thing that has me possibly concerned about this approach (renaming properties in the schema & uiSchema) is how that may trickle down to the `idSchema` and be reflected in the final submission to the API.

Have you checked the form state in the redux store to make sure this doesn't cause any problems with form submission? :thinking: We might could try submitting a form and seeing if it throws any validation errors. At any rate, I just want to make sure the submitted form data matches the schema.

bkjohnson committed:
Update src/platform/forms-system/src/js/helpers.js

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson committed:
Revert some of the changes

Shifting towards the idSChema customization

bkjohnson commented:
I'm changing approaches a bit - going to rely on changes made to the `react-jsonschema-form` dependency instead: https://github.com/department-of-veterans-affairs/react-jsonschema-form/pull/7

bkjohnson committed:
Use the right variable

bkjohnson committed:
Merge branch 'master' into 12747-a11y-id-fix

bkjohnson committed:
Remove unused helper function & tests

bkjohnson committed:
Remove getNumeratedItemSchema helper

bkjohnson committed:
Remove numerateKeys helper

bkjohnson committed:
Create idSchema in ArrayField and pass it down

bkjohnson committed:
Don't use the idSuffix param in toIdSchema

I don't think this is actually necessary since we can just manually
append the index to the id before we make the call

bkjohnson committed:
Revert "Don't use the idSuffix param in toIdSchema"

This reverts commit 03bb1880210d2b0b83e47700ebdbd7421e1c0f43.

bkjohnson committed:
Update react-jsonschema-form version

This gives us access to the version of `toIdSchema` that accepts a
suffix, and also the `<Form>` component that lets us pass an
`idSchema` prop

bkjohnson committed:
Fix failing tests

bkjohnson committed:
Fix the failing tests in a better way

bkjohnson commented on src/platform/forms-system/src/js/review/ArrayField.jsx:
```diff
@@ -226,6 +225,13 @@ class ArrayField extends React.Component {
             const itemSchema = this.getItemSchema(index);
             const itemTitle = itemSchema ? itemSchema.title : '';
 
+            const idSchema = toIdSchema(
+              itemSchema,
+              itemSchema?.$id,
+              this.props.schema.definitions,
+              index,
+            );
```
This creates an `idSchema`, and the the `index` arg is used as a suffix that gets appended to each generated id.

Next, we pass this as a prop to `<SchemaForm>`, which passes it as a prop to `<Form>` in the `react-jsonschema-form` library.
